source/shared/cpp/ObjectModel/MarkDownBlockParser.cpp

```diff
@@ -451,45 +473,133 @@ void LinkParser::CaptureLinkToken()
+// before calling this method
+// this method will return true, after it mataches new line char
+// at least once.
+bool ListParser::MatchNewBlock(std::stringstream &stream)
+{
+    if (IsNewLine(stream.peek()))
+    {
+        do
+        {
+            stream.get();
+        } while (IsNewLine(stream.peek()));
+
+        return true;
+    }
+
+    return false;
+}
+
+// ordered list marker has form of ^\d+\.\s* or [\r,\n]\d+\.\s*, and this method checks the syntax
+// this method matches \d+\.
+bool ListParser::MatchNewOrderedListItem(std::stringstream &stream, std::string &number_string)
+{
+    do
+    {
+        number_string += stream.get();
+    } while (isdigit(stream.peek()));
+
+    if (stream.peek() == '.')
+    {
+        // ordered list syntax check complete
+        stream.unget();
+        return true;
+    }
+
+    return false;
+}
+
+// parse blocks that wasn't captured
+// if what we encounter is one of following items, start of new list, list item, or new block element,
+// we do not include in the current block, we return, and have it handled by the caller
+void ListParser::ParseSubBlocks(std::stringstream &stream)
+{
+    while (!stream.eof()) 
+    { 
+        if (IsNewLine(stream.peek()))
+        {
+            int newLineChar = stream.get();
+            // check if it is the start of new block items
+            if (isdigit(stream.peek()))
+            {
+                std::string number_string = "";
+                if (MatchNewOrderedListItem(stream, number_string))
+                {
+                    break;
+                }
+                else
+                { 
+                    m_parsedResult.AddNewTokenToParsedResult(number_string);
+                }
+            }
+            else if (MatchNewListItem(stream) || MatchNewBlock(stream))
+            {
+                break;
+            }
+
+            m_parsedResult.AddNewTokenToParsedResult(newLineChar);
+        }
+        ParseBlock(stream);
+    }
+}
+// list marker has a form of ^-\s+ or [\r, \n]-\s+, and this method checks the syntax
 void ListParser::Match(std::stringstream &stream) 
 { 
-    if (stream.peek() == '-')
+    // check for the mandatory space
+    if (IsHyphen(stream.peek()))
     {
         stream.get();
-        // check for the mendatory space
         if (stream.peek() == ' ')
         {
-            // at this point, syntax check is complete, 
-            // thus any other spaces are ignored
-            while (stream.peek() == ' ')
+            // remove space
+            do
             {
                 stream.get();
-            }
-            // parse block that follows
+            } while (stream.peek() == ' ');
+
+            // at this point, syntax check is complete, 
+            // thus any other spaces are ignored
             ParseBlock(stream);
+            // parse blocks that follows
+            ParseSubBlocks(stream);
             // capture list token
             CaptureListToken();
         }
```

Comment:
Can we pull some part of this out so we can share it with the ordered list version of this function instead of duplicating the code?

Reply:
sure, i will see if what i can do.

---
In reply to: [158165495](https://github.com/Microsoft/AdaptiveCards/pull/1053#discussion_r158165495) [](ancestors = 158165495)